File: app/src/main/java/com/bartovapps/gpstriprec/timer/TimerManager.java
//package com.bartovapps.gpstriprec.timer;
//
//import android.content.Context;
//import android.os.Handler;
//import android.widget.TextView;
//
//import com.bartovapps.gpstriprec.displayers.HmsDisplayer;
//import com.bartovapps.gpstriprec.displayers.TimeDisplayer;
//
//public class com.bartovapps.gpstriprec.core.timer.TimerManager {
//
//	Context context;
//	TextView tv;
//
//	private long startTime = 0;
//	private long millis = 0;
//	private int minutes = 0;
//	private int seconds = 0;
//
//	private long pausedMsec = 0;
//	TimeDisplayer timeDisplayer;
//
//	Handler timerHandler = new Handler();
//	Runnable timerRunnable = new Runnable() {
//
//		@Override
//		public void run() {
//			millis = System.currentTimeMillis() - startTime;
//			timeDisplayer.displayTime(tv, millis);
//			timerHandler.postDelayed(this, 1000);
//		}
//	};
//
//	public com.bartovapps.gpstriprec.core.timer.TimerManager(Context context, TextView tv){
//		this.context = context;
//		this.tv = tv;
//		timeDisplayer = new HmsDisplayer();
//	}
//
//    public com.bartovapps.gpstriprec.core.timer.TimerManager(Context context){
//        this.context = context;
//    }
//
//	public void startTimer() {
//		startTime = System.currentTimeMillis();
//		timerHandler.postDelayed(timerRunnable, 0);
//	}
//
//	public void stopTimer(){
//		timerHandler.removeCallbacks(timerRunnable);
//	}
//
//	public void pauseTimer(){
//		timerHandler.removeCallbacks(timerRunnable);
//		pausedMsec = millis;
//	//	startTime = System.currentTimeMillis();
//	}
//
//	public void resetTimer(){
//		startTime = System.currentTimeMillis();
//	}
//
//	public void resumeTimer(){
//		timerHandler.postDelayed(timerRunnable, 0);
//		startTime = System.currentTimeMillis() - pausedMsec;
//	}
//
//	public double getTimeSec(){
//		return (double)(minutes * 60 + seconds);
//	}
//
//	public long getTimeMillis(){
//		return millis;
//	}
//
//	public void setStartTime(long startTime){
//		this.millis = startTime;
//		this.pausedMsec = startTime;
//		resumeTimer();
//		pauseTimer();
//	}
//
//
//
//}

File: app/src/main/kotlin/core/timer/TimerManager.kt
package com.bartovapps.gpstriprec.core.timer

import android.os.Handler
import com.bartovapps.gpstriprec.core.di.QTimerThread
import com.bartovapps.gpstriprec.presentation.displayers.TimeDisplayer
import javax.inject.Inject

interface TripTimer {
    fun startTimer()
    fun stopTimer()
    fun pauseTimer()
    fun resetTimer()
    fun resumeTimer()
    fun setStartTime(timer: Long)
    fun subscribeTimerChanges()
    val timeMillis : Long
}

class TimerManager @Inject constructor(
    @QTimerThread private val timerHandler: Handler,
    private val timeDisplayer: TimeDisplayer,
) : TripTimer {
    private var startTime: Long = 0

    private val minutes = 0
    private val seconds = 0

    private var pausedMsec: Long = 0

    override var timeMillis: Long = 0

    private val timerRunnable = Runnable {
        timeMillis = System.currentTimeMillis() - startTime
        //todo post timer value
        postTimerEvent(1000)
    }

    override fun startTimer() {
        startTime = System.currentTimeMillis()
        timerHandler.postDelayed(timerRunnable, 0)
    }

    override fun stopTimer() {
        timerHandler.removeCallbacks(timerRunnable)
    }

    override fun pauseTimer() {
        timerHandler.removeCallbacks(timerRunnable)
        pausedMsec = timeMillis
    }

    override fun resetTimer() {
        startTime = System.currentTimeMillis()
    }

    override fun resumeTimer() {
        postTimerEvent(0)
        startTime = System.currentTimeMillis() - pausedMsec
    }


    private fun postTimerEvent(millis: Long) {
        timerHandler.postDelayed(timerRunnable, millis)
    }

    val timeSec: Double
        get() = (minutes * 60 + seconds).toDouble()

    override fun setStartTime(startTime: Long) {
        this.timeMillis = startTime
        this.pausedMsec = startTime
        resumeTimer()
        pauseTimer()
    }

    override fun subscribeTimerChanges() {
        TODO("Not yet implemented")
    }
}